enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,419 @@
+
+### Goals
+
+- Prepare openshift components to accomodate the out-of-tree plugins and remove support for the in-tree implementation.
+- Describe an approach how an `external` cloud provider configuration for the cloud will be supported.
+- Provide the means to allow generic `CCM` component registration via the provider specific upstream implemetation.
+- Concider support for running switching cloud configurations in a single cluster.
+
+### Non-Goals
+
+- Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.
+- Add support for CSI and other out-of-tree implementations of cluster storage interfaces.
+
+## Proposal
+
+Add an out-of-tree support based on current cluster infrastructure, and make a seamless transition for all currently supported providers.
+
+This change will be gradually executed on all suported cloud providers: OpenStack, AWS, Azure, GCP and vSphere. The change could help inclusion for other cloud providers, such as [Yandex](https://github.com/flant/yandex-cloud-controller-manager) and [IBM](https://github.com/kubernetes/enhancements/issues/671) `CCMs` (`cloud-controller-managers`), if this will be concidered a goal.
+
+### User Stories
+
+#### Story 1:
+
+As a cloud developer, I’d like to improve support for cloud features for openshift. I'd like to develop and test new cloud providers with do not ship with default kubernetes distiribution, and require support for external cloud controller manager.
+
+#### Story 2:
+
+As a developer I'd like to develop, build and release fixes independently from the kubernetes core, and assume they will first land upstream and then will be carried over into openshift destribution with less effort.
+
+#### Story 3:
+
+We’d like to discuss technical details related to a specific cloud in a SIG meeting with people who are also involved into development in this domain, and that way gain useful insights into the cloud infrastructure, improve the overall quality of our features, stay on top of the new features, and improve the relations with maintainers outside of our company, which nevertheless share with us a common goal.
+
+### Implementation Details
+
+The `cloud-controller-manager-operator` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. That repository will manage `CCM` provisioning for metioned cloud providers.
```

Comment:
Are we going with `cluster-cloud-controller-manager-operator` for the name? If so we should update this reference.

```suggestion
The `cluster-cloud-controller-manager-operator` implementation will be hosted within openshift repository in https://github.com/openshift/cluster-cloud-controller-manager-operator. The operator will manage `CCM` provisioning for supported cloud providers.
```